src/engine/QueryPlanner.cpp

```diff
@@ -1583,13 +1589,122 @@ std::vector<SubtreePlan> QueryPlanner::runGreedyPlanningOnConnectedComponent(
+    const auto& leftChild = *children[0];
+
+    // Right child must be constant
+    auto rightChild = children[1].get();
+    auto literalExpr =
+        dynamic_cast<sparqlExpression::detail::LiteralExpression<ValueId>*>(
+            rightChild);
+    if (literalExpr == nullptr) {
+      return std::nullopt;
+    }
+    ValueId constant = literalExpr->value();
+    // TODO<ullingerc> After merge of #1938 handle distance unit
+    size_t maxDist = 0;
+    if (constant.getDatatype() == Datatype::Double) {
+      maxDist = static_cast<size_t>(round(constant.getDouble() * 1000));
+    } else if (constant.getDatatype() == Datatype::Int) {
+      maxDist = constant.getInt() * 1000;
+    } else {
+      return std::nullopt;
+    }
+
+    auto geoFuncCall = getGeoFunctionExpressionParameters(leftChild);
+    if (!geoFuncCall.has_value()) {
+      return std::nullopt;
+    }
+
+    return std::pair<sparqlExpression::GeoFunctionCall, size_t>{
+        geoFuncCall.value(), maxDist};
+  };
+
+  // Check if the filter expression is suitable for the optimization
+  for (size_t i = 0; i < filters.size(); i++) {
+    auto& filterExpression = filters[i];
+    sparqlExpression::SparqlExpression* filterExp =
+        filterExpression.expression_.getPimpl();
+    const auto& filterBody = *filterExp;
+
+    // Analyze the expression: Check if the body of the filter directly is an
+    // optimizable geof: function
+    auto geoFuncCall = getGeoFunctionExpressionParameters(filterBody);
+    std::optional<size_t> maxDist = std::nullopt;
+
+    if (!geoFuncCall.has_value()) {
+      // If the filter body is no geof: function, it can still be a maximum
+      // distance spatial search (direct body of filter is comparison).
+      auto distFilterRes = dissectWithinDistFilter(filterBody);
+      if (!distFilterRes.has_value()) {
+        continue;
+      }
+      geoFuncCall = distFilterRes.value().first;
+      maxDist = distFilterRes.value().second;
+    }
+
+    // Construct spatial join
+    auto [type, left, right] = geoFuncCall.value();
+    SpatialJoinConfiguration sjConfig{SpatialJoinConfig{type, maxDist},
+                                      left,
+                                      right,
+                                      std::nullopt,
+                                      PayloadVariables::all(),
+                                      SpatialJoinAlgorithm::LIBSPATIALJOIN,
+                                      type};
+    auto plan = makeSubtreePlan<SpatialJoin>(_qec, sjConfig, std::nullopt,
+                                             std::nullopt);
+
+    // Mark that this subtree plan (even if it does not contain a filter
+    // operation) handles the filter
+    plan._idsOfIncludedFilters |= 1ull << i;
+
+    // TODO<ullingerc> Remove this very bad workaround for assertion in
+    // findUniqueNodeIds.
+    plan._idsOfIncludedNodes = 1ull << 63;
+
+    plans.push_back(std::move(plan));
+  }
+
+  return plans;
+};
```

Comment:
Actually one question:
Do we ahve an assertion somewhere that a spatial join has to be between otherwise connected variables?
If so this would be broken by this PR. 
We should handle it more gracefully
(It should first do the spatial join and then the (possibly multicolumn) join with whatever the rest ist.

Reply:
This can be fixed by making the spatial join aware of it actually being automatically added as a substitute for a filter, and then we can silently ignore this and return just no plan here.